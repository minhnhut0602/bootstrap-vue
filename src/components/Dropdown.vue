<template>
    <div class="dropdown">
        <Btn
            class="dropdown-toggle"
            data-bs-toggle="dropdown"
            :size="size"
            :variant="variant"
        >
            {{ text }}
        </Btn>

        <DropdownMenu
            :items="items"
            @click="clickItem($event)"
        />
    </div>
</template>

<script lang="ts" setup>
import {Item, dropdownItemsProps} from '@/composables/useDropdownItems';
import {defineEmits, defineProps} from 'vue';
import Btn from '@/components/Button.vue';
import DropdownMenu from '@/components/DropdownMenu.vue';
import {sizeProps} from '@/composables/useSize';
import {variantProps} from '@/composables/useVariant';

defineProps({
    ...dropdownItemsProps,
    ...sizeProps,
    text: {
        type: String,
        required: true,
    },
    ...variantProps,
});

const emit = defineEmits(['click:item']);

const clickItem = (item: Item) => emit('click:item', item);
</script>
