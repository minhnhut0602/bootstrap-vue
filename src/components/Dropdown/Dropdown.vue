<template>
    <div class="dropdown">
        <Btn
            class="dropdown-toggle"
            data-bs-toggle="dropdown"
            :size="size"
            :variant="variant"
        >
            <slot />
        </Btn>

        <DropdownMenu
            :items="items"
            @click:item="clickItem($event)"
        />
    </div>
</template>

<script lang="ts" setup>
import {Item, dropdownItemsProps} from '../../composables/useDropdownItems';
import Btn from '../Button';
import DropdownMenu from '../DropdownMenu';
import {sizeProps} from '../../composables/useSize';
import {variantProps} from '../../composables/useVariant';

defineProps({
    ...dropdownItemsProps,
    ...sizeProps,
    ...variantProps,
});

const emit = defineEmits(['click:item']);

const clickItem = (item: Item) => emit('click:item', item);
</script>
