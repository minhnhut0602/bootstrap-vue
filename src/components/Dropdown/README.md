# Dropdown

```vue live
<Dropdown>Default Example Usage</Dropdown>
```

## Slots

| Name           | Description | Bindings |
| -------------- | ----------- | -------- |
| dropdownToggle |             |          |
| default        |             |          |
| dropdownMenu   |             | <br/>    |
| prepend        |             |          |
| items          |             |          |
| item           |             |          |
| append         |             |          |
