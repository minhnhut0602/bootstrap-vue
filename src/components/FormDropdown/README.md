# FormDropdown

```vue live
<FormDropdown>Default Example Usage</FormDropdown>
```

## Props

| Prop name   | Description | Type         | Values | Default   |
| ----------- | ----------- | ------------ | ------ | --------- |
| disabled    |             | disabledProp | -      |           |
| labelKey    |             | string       | -      | 'label'   |
| menuClass   |             | string       | -      | 'w-100'   |
| modelValue  |             | object       | -      | undefined |
| placeholder |             | string       | -      | 'Select'  |
| size        |             | sizeProp     | -      |           |
| toggleClass |             | string       | -      | undefined |

## Slots

| Name           | Description | Bindings |
| -------------- | ----------- | -------- |
| dropdownToggle |             |          |
| noOptions      |             |          |
| prepend        |             |          |
| items          |             |          |
| item           |             |          |
| append         |             |          |

## Events

| Event name        | Properties | Description |
| ----------------- | ---------- | ----------- |
| update:modelValue |            |
