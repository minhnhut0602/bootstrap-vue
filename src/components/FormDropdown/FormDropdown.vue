<template>
    <Dropdown
        class="form-dropdown"
        v-bind="propsDropdown"
        :items="filteredItems"
        @click:item="emit('update:modelValue', $event)"
    >
        <template #dropdownToggle>
            <FormControl
                v-if="search"
                v-model="searchQuery"
                :size="size"
                data-bs-toggle="dropdown"
            />

            <div
                v-else
                class="form-select"
                :class="classes"
                data-bs-toggle="dropdown"
            >
                {{ label }}
            </div>
        </template>

        <template #menuPrepend>
            <slot
                v-if="filteredItems.length === 0"
                name="noItems"
            >
                <DropdownMenuItem
                    class="pe-none"
                    label="No options"
                />
            </slot>

            <slot name="menuPrepend" />
        </template>

        <template #menuAppend>
            <slot name="menuAppend" />
        </template>
    </Dropdown>
</template>

<script lang="ts">
import useFormSelect, {formSelectProps} from '@/composables/useFormSelect';
import {computed, ref, watch} from 'vue';
import Dropdown from '@/components/Dropdown';
import DropdownMenuItem from '@/components/DropdownMenuItem';
import FormControl from '@/components/FormControl';
import {dropdownProps} from '@/composables/useDropdown';
import useClasses from '@/composables/useClasses';

const {disabled: disabledProp, size: sizeProp} = formSelectProps;
</script>

<script lang="ts" setup>
const props = defineProps({
    ...dropdownProps,
    disabled: disabledProp,
    labelKey: {
        type: String,
        default: 'label',
    },
    menuClass: {
        type: String,
        default: 'w-100',
    },
    modelValue: {
        type: Object,
        default: undefined,
    },
    placeholder: {
        type: String,
        default: 'Select',
    },
    search: {
        type: Boolean,
        default: false,
    },
    searchItems: {
        type: Boolean,
        default: true,
    },
    searchQuery: {
        type: String,
        default: '',
    },
    size: sizeProp,
    toggleClass: {
        type: String,
        default: undefined,
    },
});

const emit = defineEmits(['update:modelValue', 'update:searchQuery']);

const {classes} = useClasses(computed(() => [
    ...useFormSelect(props).classes.value,
    props.toggleClass,
]));

const filteredItems = computed(() => {
    if (searchQuery.value === '' || !props.searchItems) {
        return props.items;
    }

    return props.items
        .filter(i =>
            i.label
                .toLowerCase()
                .includes(searchQuery.value.toLocaleLowerCase()),
        );
});

const label = computed(() => {
    if (!props.modelValue) {
        return props.placeholder;
    }

    return props.modelValue[props.labelKey];
});

const propsDropdown = computed(() => {
    const properties = {};

    Object.keys(dropdownProps)
        .forEach(k => properties[k] = props[k]);

    return properties;
});

const searchQuery = ref(props.searchQuery);

// Update the `searchQuery` once the label of the value changes. This happens
// for example when a user clicks on an item. The input should show the label
// of the clicked item.
watch(
    () => props.modelValue,
    v => props.search ? (searchQuery.value = v?.[props.labelKey] || '') : false,
);

watch(
    () => props.searchQuery,
    s => props.search ? (searchQuery.value = s) : false,
);

watch(
    () => searchQuery.value,
    s => {
        if (!props.search) {
            return;
        }

        if (props.modelValue !== undefined) {
            // Prevent emitting new `update:searchQuery` events when the
            // `searchQuery` is equal to the label of the `modelValue`.
            if (props.modelValue[props.labelKey] === s) {
                return;
            }

            emit('update:modelValue', undefined);
        }

        emit('update:searchQuery', s)
    },
);
</script>
