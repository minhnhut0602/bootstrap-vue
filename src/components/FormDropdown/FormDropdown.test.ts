import {mount, shallowMount} from '@vue/test-utils';
import Dropdown from '@/components/Dropdown';
import FormDropdown from '.';

// const

describe('template', () => {
    componentRenderTest(FormDropdown, {}, false);

    it('emits update:modelValue on click item', () => {
        const wrapper = shallowMount(FormDropdown);

        wrapper.findComponent(Dropdown).vm.$emit('click:item', 'yooo');

        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['yooo']);
    });

    it('adds disabled class to .form-select', () => {
        const wrapper = mount(FormDropdown, {props: {disabled: true}});

        const formSelect = wrapper.find('.form-select');

        expect(formSelect.classes()).toContain('disabled');
    });

    it('adds size class to .form-select', () => {
        const wrapper = mount(FormDropdown, {props: {size: 'lg'}});

        const formSelect = wrapper.find('.form-select');

        expect(formSelect.classes()).toContain('form-select-lg');
    });
});

describe('label', () => {
    it('uses placeholder when no modelValue', () => {
        const wrapper = shallowMount(FormDropdown, {
            props: {placeholder: 'I am a placeholder'},
        });

        expect(wrapper.vm.label).toBe('I am a placeholder');
    });

    it('uses labelKey of modelValue when modelValue is given', () => {
        const wrapper = shallowMount(FormDropdown, {
            props: {
                modelValue: {name: 'Dees Dohmen'},
                labelKey: 'name',
            },
        });

        expect(wrapper.vm.label).toBe('Dees Dohmen');
    });
});
