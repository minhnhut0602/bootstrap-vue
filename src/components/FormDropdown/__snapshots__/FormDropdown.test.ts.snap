// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div class=\\"dropdown form-dropdown\\">
  <div class=\\"form-select form-select-md\\" data-bs-toggle=\\"dropdown\\">Select</div>
  <ul class=\\"dropdown-menu w-100\\"></ul>
</div>"
`;
